### PR TITLE
ToCode Solution: תרגול קבצים

### PR DESCRIPTION
**Pull request by EyalAbel**

### CODE REVIEW

File: 08_files_lab/01.py [hunk 1]
```diff
@@ -1,6 +1,42 @@
...
+########################################################################################################################
+#
+#
+def main():
+    parser = argparse.ArgumentParser(description='This program concatinate file_1 into file_2')
+    
+    parser.add_argument("file_1", help="path to file")
+    parser.add_argument("file_2", help="path to file")    
+    args = parser.parse_args()
+    if not is_valid_file_path(args.file_1) or not is_valid_file_path(args.file_2):
```
Comment (ynonp) [inline 08_files_lab/01.py]: Prefer using try/except over if for detecting errors. This saves you some extra tests and usually results in more correct code. For example - consider what happens if file_2 is not writable.

try/except gives you one place to handle everything that can break, and allows handling different errors differently

File: 08_files_lab/01.py [hunk 2]
```diff
@@ -1,6 +1,42 @@
...
+        sys.exit()
+
+    with open(args.file_1, "r") as src:
+        with open(args.file_2, "a") as dst:
+            dst.write(src.read())
```
Comment (ynonp) [inline 08_files_lab/01.py]: What if file_1 is too large to fit in memory?

File: 08_files_lab/02.py [hunk 3]
```diff
@@ -1,7 +1,63 @@
...
+    parser.add_argument("file_1", help="path to file")
+    parser.add_argument("file_2", help="path to file")
+    parser.add_argument("file_3", help="path to file")
+    args = parser.parse_args()
+    if not is_valid_file_path(args.file_1) or not is_valid_file_path(args.file_2):
+        sys.exit()
+    if os.path.isfile(args.file_3):
+        print "Error - file",args.file_3,"alrady exist"
+        sys.exit()
+    if not create_dictionaries(args.file_3):
```
Comment (ynonp) [inline 08_files_lab/02.py]: Don't turn your exceptions to error codes - use them all the way. Consider just the line `os.makedirs(os.path.dirname(file_path))` inside a try/except block. You'll get the same result but without defining a function and without the if, resulting in much shorter and cleaner code

File: 08_files_lab/02.py [hunk 4]
```diff
@@ -1,7 +1,63 @@
...
+        sys.exit()
+    
+
+    with open(args.file_1, "r") as src1:
+        with open(args.file_2, "r") as src2:
+            with open(args.file_3, "w") as dst:        
+                file1_lines = src1.readlines()
+                file2_lines = src2.readlines()
+                file3_lines = filter(None, sum(itertools.izip_longest(file1_lines, file2_lines), ()))
+                dst.writelines(file3_lines)
```
Comment (ynonp) [inline 08_files_lab/02.py]: With `for` you don't need to read everything into memory

File: 08_files_lab/01.py (hunk 1 above)
Comment (EyalAbel) [inline 08_files_lab/01.py]: Fixed

File: 08_files_lab/01.py (hunk 2 above)
Comment (EyalAbel) [inline 08_files_lab/01.py]: Fixed

File: 08_files_lab/02.py [hunk 5]
```diff
@@ -9,22 +9,15 @@
...
-    if not os.path.exists(os.path.dirname(file_path)):
-        try:
-            os.makedirs(os.path.dirname(file_path))
-        except OSError as exc: 
-            if exc.errno != errno.EEXIST:           
-                return False
-    return True
+    dir = os.path.dirname(file_path)
+    if not os.path.exists(dir):
+        os.makedirs(os.path.dirname(file_path))
```
Comment (ynonp) [inline 08_files_lab/02.py:15]: 1. You mean create_directories right?
2. P.S. one of my favourite python3 features is:
   https://docs.python.org/3/library/os.html?highlight=makedirs#os.makedirs

File: 08_files_lab/02.py [hunk 6]
```diff
@@ -35,29 +28,28 @@ def main():
...
-        
+    try:
+        create_dictionaries(args.file_3)
 
-    
-           
+        file1_lines = load_file(args.file_1)
+        file2_lines = load_file(args.file_2)
+        with open(args.file_3, "w") as dst:
+            itr = itertools.izip_longest(file1_lines, file2_lines)
```
Comment (ynonp) [inline 08_files_lab/02.py:42]: Prefer: `line1, line2 = itertools.izip_longest(file1_lines, file2_lines)`
So you don't need the brackets

**Pull request closed without merging**